File: mytoolboxes/HeliCam/h8SurfSimple.m
clc; clear all; close all;

if ~NET.isNETSupported
    disp('Supported .NET Framework not found')
    return
end

NET.addAssembly('C4HdlCLR');
import C4HdlCLR.*

disp('*************************************************')
disp('* Matlab example h8SurfSimple based on C4HdlCLR *')
disp('*************************************************')

c4sys = heliotis.C4HandlerCLR();
c4sys.reset();
ifNo = selectInterface(c4sys);
c4if = c4sys.openInterface(ifNo);
devNo = selectDevice(c4if);
c4dev = c4if.openDevice(devNo);

disp('initialize camera: ')
initializeDevice(c4dev)


c4dev.startAcquisition(4)

% measurement loop (10 iterations)
for i = 0:10
    % trigger single measurement
    c4dev.writeString("TriggerSelector", "FrameStart");
    c4dev.executeCommand("TriggerSoftware");

    % acquire data
    c4buf = c4dev.getBuffer(10000);
    
    % look for float surface and amplitude component
    surfPart = -1;
    ampPart = -1;
    nofParts = c4buf.readInteger("ChunkPartCount");
    for j = 0:(nofParts-1)
        c4buf.writeInteger("ChunkPartSelector", j);
        partType = c4buf.readString("ChunkPartType");
        if partType == "Surface"
            surfPart = j;
        elseif partType == "Amplitude"
            ampPart = j;
        end
    end

    rawSurf = c4buf.getDataPartFloat(surfPart);
    rawSurfDim = c4buf.getPartDimension(surfPart);

    rawAmp = c4buf.getDataPartUint16(ampPart);
    rawAmpDim = c4buf.getPartDimension(ampPart);

    % Conversion of types .NET and Matlab :
    % https://ch.mathworks.com/help/compiler_sdk/dotnet/rules-for-data-conversion-between-net-and-matlab.html 
    % See types command : whos
    % rawAmp                  1x1                       8  System.UInt16[]                      
    % rawAmpDim               1x1                       8  System.Int64[]                       
    % rawSurf                 1x1                       8  System.Double[]                      
    % rawSurfDim              1x1                       8  System.Int64[]                       
    
    matSurfDim = int64(rawSurfDim);
    surfWidth = matSurfDim(1);
    surfHeight = matSurfDim(2);
    % transpose is required for correct plotting of the data
    matSurf = transpose(reshape(double(rawSurf) , surfWidth, surfHeight));
    
    matAmpDim = int64(rawAmpDim);
    ampWidth = matAmpDim(1);
    ampHeight = matAmpDim(2);
    % transpose is required for correct plotting of the data
    matAmp = transpose(reshape(uint16(rawAmp) , ampWidth, ampHeight));
    
    figure(1); imagesc(matSurf); title("Surface"); 
    figure(2); imagesc(matAmp); title("Amplitude");

    % release buffer
    c4buf.release();

    disp("Iteration " + num2str(i) + " done.");
end

c4dev.stopAcquisition()

% release resources
c4dev.release()
c4if.release()

function ifNo = selectInterface(c4sys)
    ifNo = -1;
    disp('Interfaces: ')
    nofIf = c4sys.updateInterfaceList();
    if (nofIf == 0)
        disp('No interface detected!');
        return
    end
    for i = 0:(nofIf-1)
        curIfName = c4sys.getInterfaceName(i);
        disp(num2str(i + 1) + ": " + string(curIfName));
    end
    val = input('select an interface (exit with any other value): ');
    if ((val <= 0) || (val > nofIf))
        disp('Interface selection is out of range!');
        return
    end
    ifNo = (val - 1);
end

function devNo = selectDevice(c4if)
    devNo = -1;
    disp('Device: ')
    nofDev = c4if.updateDeviceList();
    if (nofDev == 0)
        disp('No device detected!');
        return
    end
    for i = 0:(nofDev-1)
        curDevName = c4if.getDeviceName(i);
        disp(num2str(i + 1) + ": " + string(curDevName));
    end
    val = input('select a device (exit with any other value): ');
    if ((val <= 0) || (val > nofDev))
        disp('Device selection is out of range!');
        return
    end
    devNo = (val - 1);
end

function initializeDevice(c4dev)
    % simple configuration example of heliInspect H8 using internal motion control
    
    % Motion parameters in [mm] or [mm/s]
    scanPosition = -1.4;
    scanRange = 0.5;
    scanSpeed = 5.0;
    generalSpeed = 10.0;
    % exposure ration between 0.0 and 1.0 [1.0 = 100% exposure]
    exposureRatio = 1.0;
    
    % enable required components
    c4dev.writeString("ComponentSelector", "Intensity");
    c4dev.writeInteger("ComponentEnable", 0);
    c4dev.writeString("ComponentSelector", "Range");
    c4dev.writeInteger("ComponentEnable", 1);
    c4dev.writeString("ComponentSelector", "Reflectance");
    c4dev.writeInteger("ComponentEnable", 1);
    c4dev.writeString("ComponentSelector", "Phase");
    c4dev.writeInteger("ComponentEnable", 0);
    
    % enable required chunk data (optional)
    c4dev.writeInteger("ChunkModeActive", 1);
    c4dev.writeString("ChunkSelector", "PartCount");
    c4dev.writeInteger("ChunkEnable", 1);
    c4dev.writeString("ChunkSelector", "PartType");
    c4dev.writeInteger("ChunkEnable", 1);
    
    % trigger configuration
    c4dev.writeString("TriggerSelector", "RecordingStart");
    c4dev.writeString("TriggerMode", "On");
    c4dev.writeString("TriggerSource", "Stage");
    c4dev.writeString("TriggerSelector", "AcquisitionStart");
    c4dev.writeString("TriggerMode", "Off");
    c4dev.writeString("TriggerSelector", "FrameStart");
    c4dev.writeString("TriggerMode", "On");
    c4dev.writeString("TriggerSource", "Software");
    % Hint: The TriggerSource also controls the TriggerSoftware feature.
    
    % encoder configuration
    c4dev.writeString("EncoderSelector", "Camera");
    c4dev.writeInteger("EncoderInverter", 1);
    
    % motion and position configuration
    c4dev.writeFloat("ScanPosition", scanPosition);
    c4dev.writeFloat("ScanRange", scanRange);
    c4dev.writeFloat("ScanSpeed", scanSpeed);
    c4dev.writeFloat("GeneralSpeed", generalSpeed);
    
    c4dev.writeString("ScanMode", "Down");
    c4dev.executeCommand("StageInit");
    
    % additional camera and processing configuration
    c4dev.writeString("Scan3dExtractionMethod", "AcceleratedCenterOfMassIQCorrection");
    c4dev.writeString("Scan3dScalingMethod", "zTags");
    c4dev.writeString("Scan3dDistanceUnit", "um");
    
    c4dev.writeFloat("TargetVerticalSpacing", 2.5);
    c4dev.writeFloat("ExposureRatio", exposureRatio);
    
    c4dev.writeString("FPNCorrection", "AverageLastFrames");
    c4dev.writeInteger("FPNCorrectionNFrames", 8);
    c4dev.writeInteger("ExtSimpMaxHWin", 7);
    
    % illumination control (D3)
    c4dev.writeString("LightControllerSelector", "LightController0");
    c4dev.writeString("LightControllerSource", "UserOutput0");
    c4dev.writeFloat("LightBrightness", 100.0);
    % illumination control (D2)
    % LineSelector = Line2
    % LineSource = UserOutput0
    % LineInverter = true
    % switch on the illumination (D2/D3)
    c4dev.writeString("UserOutputSelector", "UserOutput0");
    c4dev.writeInteger("UserOutputValue", 1);
end

File: mytoolboxes/HeliCam/C4Utility/examples/Python/h8SurfSimple/h8SurfSimple.py
##
# \file        h8SurfSimple.py
# \author      Silvan Murer
# \date        07 April 2021
# \copyright   Heliotis, 2021
# \version     1.0.2
#
# \brief       Simple Python example based on Harvester package for heliInspect H8
import os
import numpy as np
import matplotlib.pyplot as plt

from harvesters.core import Harvester

##
# Scan for available devices and print device list
# let the user select one and open the connection to the device
# if just one device available, open it without user interaction
# \param h harvester object
# \return harvesters camera object
def selectDevice(h):
    h.update()
    print(str(len(h.device_info_list)) + " devices detected on the Network\n\n")
    cnt = 0
    for dev in h.device_info_list:
        cnt = cnt + 1
        msg = str(cnt) + ") " + dev.id_ + " (sn:" + dev.serial_number + ")"
        print(msg)
    if cnt == 1:
        selection_int = 1
    else:
        selection_str = input("Select a device (0=exit): ")
        selection_int = int(selection_str)
        if((selection_int <= 0) or (selection_int > cnt)):
            exit('No device selected - exit script')

    deviceID = h.device_info_list[selection_int-1].id_
    print('selected device:', deviceID)
    return h.create(selection_int-1)


##
# simple configuration example of heliInspect H8 using internal motion control
# \param camera harvesters camera object
def initializeDevice(camera):
    # Motion parameters in [mm] or [mm/s]
    scanPosition = -1.4
    scanRange = 0.5
    scanSpeed = 5.0
    generalSpeed = 10.0
    # exposure ration between 0.0 and 1.0 [1.0 = 100% exposure]
    exposureRatio = 1.0

    camera.remote_device.node_map.DeviceOperationMode.value = 'Topography'

    # enable required components
    camera.remote_device.node_map.ComponentSelector.value = "Intensity"
    camera.remote_device.node_map.ComponentEnable.value = False
    camera.remote_device.node_map.ComponentSelector.value = "Range"
    camera.remote_device.node_map.ComponentEnable.value = True
    camera.remote_device.node_map.ComponentSelector.value = "Reflectance"
    camera.remote_device.node_map.ComponentEnable.value = True
    camera.remote_device.node_map.ComponentSelector.value = "Phase"
    camera.remote_device.node_map.ComponentEnable.value = False

    # enable required chunk data (optional)
    camera.remote_device.node_map.ChunkModeActive.value = True
    camera.remote_device.node_map.ChunkSelector.value = "PartCount"
    camera.remote_device.node_map.ChunkEnable.value = True
    camera.remote_device.node_map.ChunkSelector.value = "PartType"
    camera.remote_device.node_map.ChunkEnable.value = True

    # trigger configuration
    camera.remote_device.node_map.TriggerSelector.value = "RecordingStart"
    camera.remote_device.node_map.TriggerMode.value = "On"
    camera.remote_device.node_map.TriggerSource.value = "Stage"
    camera.remote_device.node_map.TriggerSelector.value = "AcquisitionStart"
    camera.remote_device.node_map.TriggerMode.value = "Off"
    camera.remote_device.node_map.TriggerSelector.value = "FrameStart"
    camera.remote_device.node_map.TriggerMode.value = "On"
    camera.remote_device.node_map.TriggerSource.value = "Software"
    # Hint: The TriggerSource also controls the TriggerSoftware feature.

    # encoder configuration
    camera.remote_device.node_map.EncoderSelector.value = "Camera"
    camera.remote_device.node_map.EncoderInverter.value = True

    # motion and position configuration
    camera.remote_device.node_map.ScanPosition.value = scanPosition
    camera.remote_device.node_map.ScanRange.value = scanRange
    camera.remote_device.node_map.ScanSpeed.value = scanSpeed
    camera.remote_device.node_map.GeneralSpeed.value = generalSpeed

    camera.remote_device.node_map.ScanMode.value = "Down"
    camera.remote_device.node_map.StageInit.execute()

    # additional camera and processing configuration
    camera.remote_device.node_map.Scan3dExtractionMethod.value = "AcceleratedCenterOfMassIQCorrection"
    camera.remote_device.node_map.Scan3dScalingMethod.value = "zTags"
    camera.remote_device.node_map.Scan3dDistanceUnit.value = "um"

    camera.remote_device.node_map.TargetVerticalSpacing.value = 4.0
    camera.remote_device.node_map.ExposureRatio.value = exposureRatio

    camera.remote_device.node_map.ExtSimpMaxHWin.value = 7

    # illumination control (D3)
    camera.remote_device.node_map.LightControllerSelector.value = "LightController0"
    camera.remote_device.node_map.LightControllerSource.value = "UserOutput0"
    camera.remote_device.node_map.LightBrightness.value = 100.0
    # illumination control (D2)
    # LineSelector = Line2
    # LineSource = UserOutput0
    # LineInverter = true
    # switch on the illumination (D2/D3)
    camera.remote_device.node_map.UserOutputSelector.value = "UserOutput0"
    camera.remote_device.node_map.UserOutputValue.value = True


print('********************************************************')
print('* Configure and acquire data in "3D surface mode"')
print('********************************************************')

h = Harvester()
# read heliotis *.CTI path from environment variable
ctiFile = os.environ['DIAPHUS_GENTL64_FILE']
h.add_file(ctiFile)

camera = selectDevice(h)

try:
    print('initialize camera')
    initializeDevice(camera)
except Exception as error:
    print(error)
    exit('init failed')

scanPos = camera.remote_device.node_map.ScanPosition.value
print("Ready for measurement. Measure on position: " + str(scanPos) + "mm")
selection_str = input("y=yes else=exit: ")
if (selection_str != "y"):
    camera.destroy()
    exit("User interrupt - exit")

print('start acquisition')
camera.start()

# setup plot
gridsize = (1, 2)
fig = plt.figure()
axSurf = plt.subplot2grid(gridsize, (0, 0))
axAmp = plt.subplot2grid(gridsize, (0, 1))
fig.suptitle('Measurements')
axSurf.set_title('Surface')
imSurf = axSurf.imshow(np.ones([512, 542]))
axAmp.set_title('Amplitude')
imAmp = axAmp.imshow(np.ones([512, 542]))
plt.show(block=False)
plt.draw()
plt.pause(0.1)

# measurement loop (10 iterations)
for itr in range(0, 10):
    try:
        # trigger single measurement
        camera.remote_device.node_map.TriggerSelector.value = "FrameStart"
        camera.remote_device.node_map.TriggerSoftware.execute()

        # acquire data
        with camera.fetch(timeout=10.0) as buffer:
            # look for float surface and amplitude component
            surfPart = -1
            ampPart = -1
            nofParts = camera.remote_device.node_map.ChunkPartCount.value
            for i in range(0, nofParts):
                camera.remote_device.node_map.ChunkPartSelector.value = i
                partType = camera.remote_device.node_map.ChunkPartType.value
                if(partType == "Surface"):
                    curPixelFormat = buffer.payload.components[i].data_format
                    if(curPixelFormat == 'Mono16'):
                        pass
                    elif(curPixelFormat == 'Coord3D_C32f'):
                        surfPart = i
                elif(partType == "Amplitude"):
                    ampPart = i

            # plot surface and amplitude component
            if (surfPart >= 0):
                height = buffer.payload.components[surfPart].height
                width = buffer.payload.components[surfPart].width
                surface = buffer.payload.components[surfPart].data.reshape(height, width)
                surface = np.copy(surface)
                imSurf.set_data(surface)
                imSurf.set_clim(vmin=np.min(surface), vmax=np.max(surface))
            if (ampPart >= 0):
                height = buffer.payload.components[ampPart].height
                width = buffer.payload.components[ampPart].width
                amplitude = buffer.payload.components[ampPart].data.reshape(height, width)
                amplitude = np.copy(amplitude)
                imAmp.set_data(amplitude)
                imAmp.set_clim(vmin=np.min(amplitude), vmax=np.max(amplitude))
            plt.draw()
            plt.pause(0.1)
            print("iteration " + str(itr) + " done!\n")
    except Exception as error:
        print(error)
        exit('acquisition error')

# stop acquisition and close camera connection
camera.stop()
camera.destroy()
print('example end')
